kliff/trainer/base_trainer.py

```diff
@@ -0,0 +1,636 @@
+        repository. If KIM type model is installed in CWD, it will be loaded from there, and
+        model_path will be set to the KIM CWD. If model is of type TAR, it will be untarred
+        and model_path will be set to the untarred directory. Left for the derived classes
+        to implement.
+        """
+        raise TrainerError("setup_model not implemented.")
+
+    def setup_parameter_transforms(self):
+        """
+        This method set up the transformed parameter space for models. It can be used
+        for any model type in general, but as there exists a significant difference
+        between how models handles their parameters, it is left for the subclass to
+        implement. Although to ensure that `initialize` function remains consistent
+        this method will not raise NotImplemented error, rather it will quietly pass.
+        So be aware.
+        """
+        pass
+
+    def setup_optimizer(self):
+        """
+        Set up the optimizer based on the provided information. If the optimizer is not
+        provided, it will be loaded from the optimizer_name. If the optimizer_name is not
+        provided, it will raise an error. If the optimizer_provider is scipy, it will be
+        loaded from the scipy.optimize. If the optimizer_provider is torch, it will be
+        loaded from the torch.optim. Left for the derived classes to implement.
+        """
+        raise TrainerError("setup_optimizer not implemented.")
+
+    def setup_test_train_datasets(self):
+        """
+        Simple test train split for now, will have more options like stratification
+         in the future.
+
+        """
+        # test train splits
+        train_size = self.training_manifest.get("training_dataset", {}).get(
+            "train_size", len(self.dataset)
+        )
+        val_size = self.training_manifest.get("validation_dataset", {}).get(
+            "val_size", 0
+        )
+
+        # sanity checks
+        if not isinstance(train_size, int) or train_size < 1:
+            logger.warning(
+                "Train size is not provided or is less than 1. Using full dataset for training."
+            )
+            train_size = len(self.dataset)
+        else:
+            logger.info(f"Training dataset size: {train_size}")
+
+        if not isinstance(val_size, int) or val_size < 0:
+            logger.warning(
+                "Val size is not provided or is less than 0. Using 0 for validation."
+            )
+            val_size = 0
+        else:
+            logger.info(f"Validation dataset size: {val_size}")
+
+        if train_size + val_size > len(self.dataset):
+            raise TrainerError(
+                "Sum of train, val, and test sizes is greater than the dataset size."
+            )
+
+        # check if indices are provided
+        train_indices = self.dataset_sample_manifest.get("train_indices")
+        if train_indices is None:
+            train_indices = np.arange(train_size, dtype=int)
```

Comment:
This assigns the first `train_size` data points to the trainset, and the next `val_size` to the val set. For now, we at least need to support random split, using something like
```python
indices = np.random.permutation(trian_size + val_size)
train_indices = indices[:train_size]
val_indices = indices[train_size:]
```